storage/remote/wal_watcher.go

```diff
@@ -0,0 +1,542 @@
+	SeriesReset(int)
+	Name() string
+}
+
+// WALWatcher watches the TSDB WAL for a given WriteTo.
+type WALWatcher struct {
+	writer         writeTo
+	logger         log.Logger
+	walDir         string
+	currentSegment int
+	lastCheckpoint string
+	startTime      int64
+
+	samplesReadMetric       prometheus.Counter
+	seriesReadMetric        prometheus.Counter
+	tombstonesReadMetric    prometheus.Counter
+	invalidReadMetric       prometheus.Counter
+	unknownReadMetric       prometheus.Counter
+	recordDecodeFailsMetric prometheus.Counter
+	samplesSentPreTailing   prometheus.Counter
+	currentSegmentMetric    prometheus.Gauge
+
+	quit chan struct{}
+}
+
+// NewWALWatcher creates a new WAL watcher for a given WriteTo.
+func NewWALWatcher(logger log.Logger, writer writeTo, walDir string, startTime int64) *WALWatcher {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	w := &WALWatcher{
+		logger:    logger,
+		writer:    writer,
+		walDir:    path.Join(walDir, "wal"),
+		startTime: startTime,
+		quit:      make(chan struct{}),
+	}
+
+	w.samplesReadMetric = watcherSamplesRecordsRead.WithLabelValues(w.writer.Name())
+	w.seriesReadMetric = watcherSeriesRecordsRead.WithLabelValues(w.writer.Name())
+	w.tombstonesReadMetric = watcherTombstoneRecordsRead.WithLabelValues(w.writer.Name())
+	w.unknownReadMetric = watcherUnknownTypeRecordsRead.WithLabelValues(w.writer.Name())
+	w.invalidReadMetric = watcherInvalidRecordsRead.WithLabelValues(w.writer.Name())
+	w.recordDecodeFailsMetric = watcherRecordDecodeFails.WithLabelValues(w.writer.Name())
+	w.samplesSentPreTailing = watcherSamplesSentPreTailing.WithLabelValues(w.writer.Name())
+	w.currentSegmentMetric = watcherCurrentSegment.WithLabelValues(w.writer.Name())
+
+	return w
+}
+
+func (w *WALWatcher) Start() {
+	level.Info(w.logger).Log("msg", "starting WAL watcher", "queue", w.writer.Name())
+	go w.runWatcher()
+}
+
+func (w *WALWatcher) Stop() {
+	level.Info(w.logger).Log("msg", "stopping WAL watcher", "queue", w.writer.Name())
+	close(w.quit)
+}
+
+// TODO: fix the exit logic for this function
+// The stop param is used to stop at the end of the existing WAL on startup,
+// since scraped samples may be written to the latest segment before we finish reading it.
+func (w *WALWatcher) readSeriesRecords(r *wal.LiveReader, index int, stop int64) {
+	var (
+		dec     tsdb.RecordDecoder
+		series  []tsdb.RefSeries
+		samples []tsdb.RefSample
+		ret     bool
+	)
+	for !isClosed(w.quit) {
+		for r.Next() {
+			series = series[:0]
+			rec := r.Record()
+			// If the timestamp is > start then we should Append this sample and exit readSeriesRecords,
+			// because this is the first sample written to the WAL after the WAL watcher was started.
+			typ := dec.Type(rec)
+			if typ == tsdb.RecordSamples {
+				samples, err := dec.Samples(rec, samples[:0])
+				if err != nil {
+					continue
+				}
+				for _, s := range samples {
+					if s.T > w.startTime {
+						w.writer.Append(samples)
```

Comment:
I think I'm missing something here; is `samples` contains N samples, won't this try and append them N times?

Reply:
Yes, it should be `Append(s)` :man_facepalming: